woocommerce/woocommerce-ios
# Abstract product form view model with `ProductFormViewModelProtocol`

Prerequisite for #2085 

## Changes

- Created `ProductFormViewModelProtocol` that `ProductFormViewModel` conforms to with associated type `ProductVariationFormViewModel` being `Product` (in a future PR, there will be a `ProductVariationFormViewModel` where its `ProductModel` is `ProductVariation`)
- Updated `ProductFormViewController` with generics changes - if a class has generics (`ProductFormViewModelProtocol`'s `ProductModel` in this case), all `@objc` and OjcC protocol conformance (e.g. `UITableViewDelegate`) cannot be in the extension. Also, there cannot be nested enums under a class with generics, so `ActionSheetStrings` and `Constants` were moved to private enums and `ProductFormViewController .ProductFormPresentationStyle` enum was updated to be `ProductFormPresentationStyle`.
- In `ProductFormViewController`, updated its initializer to take in a generic `viewModel: ViewModel` and `ProductImageActionHandler` because the view model depends on it.
- In `ProductDetailsFactory`, updated how `ProductFormViewController` is initialized by passing a view model and `ProductImageActionHandler`.

## Testing

No changes in app behavior are expected, feel free to test editing a product with the following steps:

- Go to the Products tab
- Tap on a simple product
- Tap on each row, potentially edit some fields, open the bottom sheet --> everything should work as before

Update release notes:
